### DISCUSSION TITLE
How to implement `value()` helper method from Laravel in Psalm?

### DISCUSSION THREAD

**Discussion by ppmathis** [vimeo/psalm#6258] [Q&A]
Is there any way to represent a function in Psalm which acts like this?

- If first argument is `Closure():T`, call it with all other arguments passed and return its return value with type `T`
- If first argument is anything else than `Closure():T`, return the value directly

The goal at the end of the day is to write a proper stub for the `value()` helper of Laravel which does exactly that and is often used to allow default arguments of other helpers/functions to either contain a value directly or a Closure returning the desired value.

I already tried https://psalm.dev/r/8ee69ae96b and would have imagined that this works, but for some reason Psalm returns `impure-Closure():int|int` for the last variable instead of just `int` as expected.

I noticed that narrowing `@template T` to `@template T as scalar|null` fixes the issue, but this is not really a solution as `value` is not just limited to scalars either. It seems to me that Psalm thinks a `Closure():int` could either match `Closure(...):T` or `T` and does not infer correctly that as soon as `T` is a `Closure()`, it can never be directly returned.

**Comment by weirdan** [comment]
You can get this with a conditional return type and another template parameter for closure return type: https://psalm.dev/r/3d49052546

**Reply by ppmathis** [reply]
Thank you! I tried various combinations of template and conditional types but never came up with this solution. I now tried integrating this into another method with a similar signature which will internally call `value()` and while all variables `a` to `e` have the correct type, Psalm complaints about `InvalidReturnStatement` and `InvalidReturnType`: https://psalm.dev/r/6651aeed3d

Is there any way to get this working without adding `/** @var int|null */` above the `return value($default);` statement?
